things/service/src/main/java/org/eclipse/ditto/things/service/persistence/actors/strategies/commands/MigrateThingDefinitionStrategy.java

```diff
@@ -0,0 +1,279 @@
+import org.eclipse.ditto.rql.query.filter.QueryFilterCriteriaFactory;
+import org.eclipse.ditto.rql.query.things.ThingPredicateVisitor;
+import org.eclipse.ditto.things.model.Thing;
+import org.eclipse.ditto.things.model.ThingId;
+import org.eclipse.ditto.things.model.ThingsModelFactory;
+import org.eclipse.ditto.things.model.signals.commands.ThingCommandSizeValidator;
+import org.eclipse.ditto.things.model.signals.commands.ThingResourceMapper;
+import org.eclipse.ditto.things.model.signals.commands.modify.MergeThing;
+import org.eclipse.ditto.things.model.signals.commands.modify.MigrateThingDefinition;
+import org.eclipse.ditto.things.model.signals.events.ThingEvent;
+
+
+/**
+ * Strategy to handle the {@link MigrateThingDefinition} command.
+ *
+ * This strategy processes updates to a Thing's definition, applying necessary data migrations
+ * and ensuring that defaults are properly initialized when required.
+ *
+ * Assumptions:
+ * - The {@link MigrateThingDefinition} command provides a ThingDefinition URL, which is used
+ *   to create a skeleton Thing. The command's payload also includes migration data and patch
+ *   conditions for fine-grained updates.
+ * - Patch conditions are evaluated using RQL-based expressions to determine which migration
+ *   payload entries should be applied.
+ * - Skeleton generation extracts and merges Thing definitions and default values separately,
+ *   ensuring a clear distinction between structural updates and default settings.
+ * - After applying skeleton-based modifications and migration payloads, the changes are merged
+ * - Property initialization can be optionally enabled via the command, applying default values
+ *   to the updated Thing when set to true.
+ * - The resulting Thing undergoes validation to ensure compliance with WoT model constraints
+ *   before persisting changes.
+ */
+@Immutable
+public final class MigrateThingDefinitionStrategy extends AbstractThingModifyCommandStrategy<MigrateThingDefinition> {
+
+    private static final ThingResourceMapper<Thing, Optional<EntityTag>> ENTITY_TAG_MAPPER =
+            ThingResourceMapper.from(EntityTagCalculator.getInstance());
+
+    private static final TimePlaceholder TIME_PLACEHOLDER = TimePlaceholder.getInstance();
+
+
+    public MigrateThingDefinitionStrategy(final ActorSystem actorSystem) {
+        super(MigrateThingDefinition.class, actorSystem);
+    }
+
+    @Override
+    protected Result<ThingEvent<?>> doApply(final Context<ThingId> context,
+            @Nullable final Thing thing,
+            final long nextRevision,
+            final MigrateThingDefinition command,
+            @Nullable final Metadata metadata) {
+
+        final Thing existingThing = getEntityOrThrow(thing);
+        final Instant eventTs = getEventTimestamp();
+
+        return handleUpdateDefinition(context, existingThing, eventTs, nextRevision, command, metadata)
+                .toCompletableFuture()
+                .join();
```

Comment:
No joining completableFutures like this - this will block the main actor thread and could cause death of the whole actor system.
There are APIs at `ResultFactory.newMutationResult` which support completionStages instead of sync API.